### DISCUSSION TITLE
Make plugin loading clear

### DISCUSSION THREAD

**Discussion by SkeLLLa** [verdaccio/verdaccio#1357] [Ideas]
Nowadays plugin loading is kind of "magic" that's hidden deep inside the code of verdaccio. 

Plugin installation (especially in dockerized env) does not work. If you install plugin globally verdaccio doesn't see it. If you try to install it into verdaccio folder for some reason after that it couldn't load `lodash`, `express`, etc.

Also it's not possible or rather hard to load some forked plugin.

Proposition:
Make a config section where plugins can be directly configured with path to them, etc.

**Comment by DanielRuf** [comment]
Hi @SkeLLLa,

What exactly do you mean with the following?

> Make a config section where plugins can be directly configured with path to them, etc.

**Comment by juanpicado** [comment]
I'm also curious to know which issues are you facing using plugins with Docker images. However, we have working examples here https://github.com/verdaccio/docker-examples

**Comment by SkeLLLa** [comment]
@DanielRuf I mean to have some kind of dictionary where users can specify what plugins they want to load and from where to load them.
For example, let's say I've added some customization into some existing verdaccio plugin (verdaccio-audit for example). And I've decided to name it `foobar`. So what I'm talking about is:

```json5
{
  plugins: [
    {name: 'verdaccio-audit': dir: '/home/user/foobar'}
  }
}
```
So here we'll load some code from `/home/user/foobar` directory and use it as 'verdaccio-audit' module.
Also if for some reason we have some other plugins in `node_modules` or elsewhere verdaccio will not use them, until they will be specified in plugins section.
If no `dir` specified it may look for this module in default location (e.g. in `node_modules`)

@juanpicado There's only one way when all's up and running. So when you're using basic docker image of verdaccio you'll need:
1) Install dev dependencies via `npm i`.
2) Install plugins **only** to verdaccio working directroy. (e.g. `npm i verdaccio-audit`). Global installation doesn't work., installing into`/verdaccio/plugins` directory may not work in some cases.
3) Afterwards run `npm run code:docker-build` again. Elsewere you'll get a lot of weird module not found errors.
4) Optionally remove dev dependencies `npm prune production`.

**Comment by juanpicado** [comment]
> Install plugins only to verdaccio working directroy. (e.g. npm i verdaccio-audit). Global installation doesn't work., installing into/verdaccio/plugins directory may not work in some cases.

Thanks for the reply, but I'm still confused about the reason of this request, does the following example does not work for you?  

// Dockerfile
```docker
FROM verdaccio/verdaccio:3

RUN npm i && npm install verdaccio-xxx-plugin-xxx-name
```
That extends the base image and you are able to install globally other packages. 

Why modify the verdaccio `Dockerfile`  directly? 🤔 I'd be great to know more about your approach.
